### PR TITLE
Fix helper paths for ceil/floor in asm.js

### PR DESCRIPTION
**Pull request by MikeHolman**
asm.js ceil/floor instructions are JIT incorrectly on x86 machines with SSE2 but without SSE4.1. There was previously a path for this, but only implemented on x64.
I have removed architecture specific nature and moved to use more standard APIs for emitting the call. I also added various asserts to catch this type of issue when using -sse switches without needing specific machine hardware.

[Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/40)

### CODE REVIEW

**Comment by dilijev** [comment]
@dotnet-bot test this please

**Comment by Cellule** [comment]
Reviewed 7 of 11 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[lib/Backend/LowerMDShared.cpp, line 1523 \[r2\]](https://reviewable.io:443/reviews/microsoft/chakracore/40#-K81FLQh2Vaxfy9sPmI5:-K81FLQh2Vaxfy9sPmI6:-200783806) ([raw file](https://github.com/microsoft/chakracore/blob/4ef410ed719132c33fd54c46bc91df55caa7783c/lib/Backend/LowerMDShared.cpp#L1523)):*
I would put a comment here like `// Start of SSE2 instructions` to avoid wrong instruction between `ADDSD` and `SUBPD`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/40)*

**Comment by talibanes** [comment]
De verdad?

**Comment by abchatra** [comment]
:+1:

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -597,7 +597,7 @@ Lowerer::LowerRange(IR::Instr *instrStart, IR::Instr *instrEnd, bool defaultDoFa
             break;
 
         case Js::OpCode::InlineMathFloor:
-#if _M_X64
+#ifdef ASMJS_PLAT
```
Comment (LouisLaf) [inline lib/Backend/Lower.cpp]: ASMJS_PLAT isn't the appropriate #define here as this is platform specific code.  Should be (x86 || x64) && ASMJS_PLAT.

File: lib/Backend/LowerMDShared.cpp [hunk 2]
```diff
@@ -1512,22 +1521,25 @@ LowererMD::Legalize(IR::Instr *const instr, bool fPostRegAlloc)
...
         case Js::OpCode::ADDPD:
         case Js::OpCode::SUBSD:
+        case Js::OpCode::ANDPD:
+        case Js::OpCode::ANDNPD:
+        case Js::OpCode::DIVPD:
+        case Js::OpCode::MAXPD:
+        case Js::OpCode::MINPD:
+        case Js::OpCode::MULPD:
+        case Js::OpCode::SUBPD:
+            Assert(AutoSystemInfo::Data.SSE2Available());
```
Comment (LouisLaf) [inline lib/Backend/LowerMDShared.cpp]: Maybe these asserts should be in DbCheckLower?  We should add a machine dependent for it...

Comment (MikeHolman) [inline lib/Backend/LowerMDShared.cpp]: DbCheckLower calls this to validate encoding, so I thought it would be better to keep that here

Comment (LouisLaf) [inline lib/Backend/LowerMDShared.cpp]: True.  Makes sense then.
Rest looks goods!

**Comment by MikeHolman** [comment]
Thanks guys! I updated the branch with couple minor changes from your comments, but GitHub won't update the PR... so closing and will reopen/merge.

**Pull request closed without merging**